File: credit.c
#include <cs50.h>
#include <stdio.h>

const int MAX_DIGITS = 16;

int main(void)
{
    int digit_count = 0;
    int digits[MAX_DIGITS];

    // Credit card numbers can start with 0,
    // however the assignment suggests to use get_long which would not capture leading zeros
    // therefore case with credit card starting with 0 is being ignored

    long number = get_long("Number: ");
    while (number > 0)
    {
        // stores the number backwards
        digits[digit_count] = (int) (number % 10);
        number = number / 10;
        digit_count += 1;
    }

    int checksum1 = 0;
    int checksum2 = 0;
    for (int i = 0; i < digit_count; i++)
    {
        if (i % 2 == 0)
        {
            checksum1 += digits[i];
        }
        else
        {
            int multiplied = digits[i] * 2;
            checksum2 += (multiplied % 10) + (multiplied / 10);
        }
    }

    if ((checksum1 + checksum2) % 10 != 0)
    {
        printf("INVALID\n");
    }
    else if (digit_count == 15 && digits[digit_count - 1] == 3 && (digits[digit_count - 2] == 4 || digits[digit_count - 2] == 7))
    {
        printf("AMEX\n");
    }
    else if (digits[digit_count - 1] == 4 && (digit_count == 13 || digit_count == 16))
    {
        printf("VISA\n");
    }
    else if (digits[digit_count - 1] == 5 &&
             (digits[digit_count - 2] == 1 || digits[digit_count - 2] == 2 || digits[digit_count - 2] == 3 ||
              digits[digit_count - 2] == 4 || digits[digit_count - 2] == 5))
    {
        printf("MASTERCARD\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

File: credit.py
import cs50

number = cs50.get_string("Number: ")

digits = list(number)

digit_count = len(digits)

digits_reversed = []
for digit in digits:
    digits_reversed.append(int(digit))
digits_reversed.reverse()

check_sum_1 = 0
check_sum_2 = 0

for i in range(0, digit_count):
    if i % 2 == 0:
        check_sum_1 += digits_reversed[i]
    else:
        multiplied = digits_reversed[i] * 2
        check_sum_2 += (multiplied % 10) + int((multiplied / 10))

digits = digits_reversed
digits.reverse()

if (check_sum_1 + check_sum_2) % 10 != 0:
    print("INVALID")
elif digit_count == 15 and digits[0] == 3 and (digits[1] in [4, 7]):
    print("AMEX")
elif digits[0] == 4 and (digit_count in [13, 16]):
    print("VISA")
elif digits[0] == 5 and digits[1] in [1, 2, 3, 4, 5]:
    print("MASTERCARD")
else:
    print("INVALID")
